Unify sbt shell and build.sbt syntax (scope path syntax)

eed3si9n opened the pull request:
Fixes sbt/sbt#1812

This adds scope path syntax (fka unified slash syntax) for both sbt shell and the build.sbt DSL.

### Scope path syntax in shell

Instead of the current `<project-id>/config:intask::key`, this **adds** `<project-id>/<config-ident>/intask/key` where `<config-ident>` is the Scala identifier notation for the configurations like `Compile` and `Test`. (The old shell syntax will continue to function)

### Scope path syntax in build.sbt

This also adds a series of implicits called `SlashSyntax` that adds `/` operators to project references, configuration, and keys such that the same syntax works in build.sbt.

### Examples

These examples work *both* from the shell *and* in build.sbt.

```scala
Global / cancelable
ThisBuild / scalaVersion
Test / test
root / Compile / compile / scalacOptions
ProjectRef(uri("file:/xxx/helloworld/"),"root")/Compile/scalacOptions
Zero / Zero / name
```

The inspect command now outputs something that can be copy-pasted:

    > inspect compile
    [info] Task: sbt.inc.Analysis
    [info] Description:
    [info] 	Compiles sources.
    [info] Provided by:
    [info] 	ProjectRef(uri("file:/xxx/helloworld/"), "root") / Compile / compile
    [info] Defined at:
    [info] 	(sbt.Defaults) Defaults.scala:326
    [info] Dependencies:
    [info] 	Compile / manipulateBytecode
    [info] 	Compile / incCompileSetup
    [info] Reverse dependencies:
    [info] 	Compile / printWarnings
    [info] 	Compile / products
    [info] 	Compile / discoveredSbtPlugins
    [info] 	Compile / discoveredMainClasses
    [info] Delegates:
    [info] 	Compile / compile
    [info] 	compile
    [info] 	ThisBuild / Compile / compile
    [info] 	ThisBuild / compile
    [info] 	Zero / Compile / compile
    [info] 	Global / compile
    [info] Related:
    [info] 	Test / compile

![slash](https://user-images.githubusercontent.com/184683/30958192-5f349404-a40a-11e7-95ed-43fc2966b4fa.gif)

jvican commented:
Two questions:
  
1. What's missing on this front for this to be merged?
2. Is there any chance this can make it into the next sbt release?

eed3si9n commented:
1. I need to add mima exclusion, I think to pass Travis.
2. I am hoping it will be in sbt 1.1.0, 1.2.0 or some feature release. Immediately next one will be sbt 1.0.2 (hotfix release).

dwijnand commented:
I've added this to the 1.1.0 milestone as it would be a good highlight feature of sbt 1.1.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n commented:
Rebased to the latest 1.x branch, and added Mima exclusion.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

dwijnand commented:
Could you squash the commits?

jvican commented on sbt/src/sbt-test/project/unified/test:
```diff
@@ -0,0 +1,13 @@
+> show ThisBuild / scalaVersion
+
+> show Test / test
+
```
`inspect`, `inspect-tree` and other commands taking keys should be tested.

jvican commented on sbt/src/sbt-test/project/unified/test:
```diff
@@ -0,0 +1,13 @@
+# "Global" now works in shell. optional whitespace arond /
+> show Global / cancelable
+
+# "ThisBuild" now works in shell. optional whitespace arond /
```
The optional whitespace around `/` should be tested for every example.
  
Also, typo in "arond".

jvican commented on sbt/src/sbt-test/project/unified/test:
```diff
@@ -0,0 +1,13 @@
+# "ThisBuild" now works in shell. optional whitespace arond /
+> show ThisBuild / scalaVersion
+
+> show Test / test
```
How do we know that we're calling the right `Test / test`? This tests that the parser accepts it and it runs something, but we don't know which something has been run.

jvican reviewed (commented):
Comments in tests.

eed3si9n committed:
Unified slash syntax

Fixes sbt/sbt#1812

This adds unified slash syntax for both sbt shell and the build.sbt DSL.
Instead of the current `<project-id>/config:intask::key`,
this adds `<project-id>/<config-ident>/intask/key` where <config-ident> is the Scala identifier notation for the configurations like `Compile` and `Test`.

This also adds a series of implicits called `SlashSyntax` that adds `/` operators to project refererences, configuration, and keys such that the same syntax works in build.sbt.

These examples work for both from the shell and in build.sbt.

    Global / cancelable
    ThisBuild / scalaVersion
    Test / test
    root / Compile / compile / scalacOptions
    ProjectRef(uri("file:/xxx/helloworld/"),"root")/Compile/scalacOptions
    Zero / Zero / name

The inspect command now outputs something that can be copy-pasted:

    > inspect compile
    [info] Task: sbt.inc.Analysis
    [info] Description:
    [info] 	Compiles sources.
    [info] Provided by:
    [info] 	ProjectRef(uri("file:/xxx/helloworld/"),"root")/Compile/compile
    [info] Defined at:
    [info] 	(sbt.Defaults) Defaults.scala:326
    [info] Dependencies:
    [info] 	Compile/manipulateBytecode
    [info] 	Compile/incCompileSetup
    [info] Reverse dependencies:
    [info] 	Compile/printWarnings
    [info] 	Compile/products
    [info] 	Compile/discoveredSbtPlugins
    [info] 	Compile/discoveredMainClasses
    [info] Delegates:
    [info] 	Compile/compile
    [info] 	compile
    [info] 	ThisBuild/Compile/compile
    [info] 	ThisBuild/compile
    [info] 	Zero/Compile/compile
    [info] 	Global/compile
    [info] Related:
    [info] 	Test/compile

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

eed3si9n commented:
@jvican @dwijnand Thanks guys for the review.

eed3si9n commented on sbt/src/sbt-test/project/unified/build.sbt:
```diff
@@ -0,0 +1,70 @@
+    libraryDependencies += uTest % Test,
+    testFrameworks += new TestFramework("utest.runner.Framework"),
+
+    commands += Command("inspectCheck", inspectBrief, inspectDetailed)(Inspect.parser) {
```
@jvican Added the inspect output test here.
Since the parser is shared, I didn't feel the need to test various modes for inspect, but the default mode has displays the keys, so I've added the test.

eed3si9n commented on sbt/src/sbt-test/project/unified/test:
```diff
@@ -0,0 +1,28 @@
+$ mkdir target
+> Test/test
+# Check the side-effect of executing uTest
+$ exists target/foo
```
@jvican Added a small "test" that should check that that uTest in fact ran.

eed3si9n commented on sbt/src/sbt-test/project/unified/test:
```diff
@@ -0,0 +1,28 @@
+> show Global / cancelable
+
+# "ThisBuild" now works in shell.
+> show ThisBuild/scalaVersion
```
@jvican Added variants with and without whitespace around /.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

dwijnand commented on main-settings/src/main/scala/sbt/Reference.scala:
```diff
@@ -91,7 +91,7 @@ object Reference {
       case LocalProject(id)    => "{<this>}" + id
       case RootProject(uri)    => "{" + uri + " }<root>"
-      case ProjectRef(uri, id) => "{" + uri + "}" + id
+      case ProjectRef(uri, id) => s"""ProjectRef(uri("$uri"),"$id")"""
```
could you add a space, please? `s"""ProjectRef(uri("$uri"), "$id")"""`

eed3si9n replied:
Traditionally `toString` doesn't put whitespaces, so I stuck to the convention:

```
sbt:sbtRoot> eval ProjectRef(uri("uri"),"id")
[info] ans: sbt.ProjectRef = ProjectRef(uri,id)
```

dwijnand replied:
It's not universally consistent:

```scala
scala> List(1, 2, 3)
res0: List[Int] = List(1, 2, 3)
```

and I can't find a single instance of creating `ProjectRef` where we _don't_ include a space: https://github.com/sbt/sbt/search?utf8=%E2%9C%93&q=ProjectRef.

Is the point not that we print what people would write?

eed3si9n replied:
I don't expect the build user to put the whitespaces around in the shell. The same goes for `/` by the way:

```
sbt:helloworld> inspect test
[info] Task: Unit
[info] Description:
[info] 	Executes all tests.
[info] Provided by:
[info] 	ProjectRef(uri("file:/Users/eed3si9n/work/hellotest/"),"root")/Test/test
[info] Defined at:
[info] 	(sbt.Defaults.testTasks) Defaults.scala:670
[info] Dependencies:
[info] 	Test/executeTests
[info] 	Test/test/streams
[info] 	Test/state
[info] 	Test/test/testResultLogger
```

eed3si9n replied:
You know what? Let me just put whitespaces to both in a separate commit. If they copy-paste to shell, shell can parse them anyway.

eed3si9n replied:
c67fe067cf63a1f18d61aaec84442348a4b09cd3

dwijnand commented on main-settings/src/main/scala/sbt/Def.scala:
```diff
@@ -57,17 +57,36 @@ object Def extends Init[Scope] with TaskMacroExtra {
+      case `current`               =>
+        // ignore (multi) and always display relative key
+        // if (multi) current.project + slash else ""
+        ""
```
Could you explain what's going on here? Why is multi ignored now?

eed3si9n replied:
`multi = true` is passed in whenever you have multi-projects. Prior to this change, when sbt detected a multi-project build it will display `$subproject/key` even though the method is called `displayRelative`. At least the Build URI is omitted, but I am making it *more* relative by omitting the subproject id.

Here's how this would show up. When you run `inspect test` in sbt 1.0.0 you get:

```
[info] Dependencies:
[info] 	sbtRoot/test:executeTests
[info] 	sbtRoot/test:test::streams
[info] 	sbtRoot/test:state
[info] 	sbtRoot/test:test::testResultLogger
```

In sbt 1.1.0 you'd get:

```
[info] Dependencies:
[info] 	Test/executeTests
[info] 	Test/test/streams
[info] 	Test/state
[info] 	Test/test/testResultLogger
```

I think this is better, because what's something the user can type in or use in `build.sbt`.

dwijnand replied:
I see. So can we lose the commented out code and the `multi` parameter from this new `private[sbt]` method?

dwijnand commented on main-settings/src/main/scala/sbt/Scope.scala:
```diff
@@ -123,29 +123,84 @@ object Scope {
+    displayMasked(scope, sep, showProject, mask, false)
+
+  /**
+   * unified / style introduced in sbt 0.13.16 / sbt 1.0
```
This wasn't introduced in 0.13.16

dwijnand commented on main-settings/src/main/scala/sbt/Scope.scala:
```diff
@@ -123,29 +123,84 @@ object Scope {
+
+  /**
+   * unified / style introduced in sbt 0.13.16 / sbt 1.0
+   * By defeault, sbt will no longer display the Zero-config,
```
default

dwijnand commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -2336,13 +2336,20 @@ object Classpaths {
+    val thisRef = thisProjectRef.value
+    val label =
+      if (isPlugin) Reference.display(thisRef)
+      else Def.displayRelativeReference(extracted.currentRef, thisRef)
```
huh, TIL that thisProjectRef != "the current project"

dwijnand commented on main/src/main/scala/sbt/SlashSyntax.scala:
```diff
@@ -0,0 +1,148 @@
+import sbt.librarymanagement.Configuration
+
+/**
+ * ScopeSyntax implements the slash syntax to scope keys for build.sbt DSL.
```
What is "ScopeSyntax"? Either it should read "SlashSyntax" or it needs more explanation.

eed3si9n replied:
ScopeSyntax was the older name. I'll change this.

dwijnand commented on main/src/main/scala/sbt/internal/KeyIndex.scala:
```diff
@@ -87,13 +109,34 @@ private[sbt] final class AKeyIndex(val data: Relation[Option[AttributeKey[_]], S
+
+/*
+ * data contains the mapping between a configuration and keys.
+ * identData contains the mapping between a configuration and its identifier.
```
why is this necessary?

eed3si9n replied:
We can't change the type of `Def.ScopeKey[_]`, which binds to `sbt.Scope`, which includes `sbt.ConfigKey`, a wrapper around a `String` that represents the `name` of configuration. The mapping information between the Scala identifier and `String` thus needs to be available during conversion back and forth. For example, if the user types in `IntegrationTest/test`, the new parser needs to parse it as `ConfigKey("it")` and `AttributeKey[Unit]("test")`.

dwijnand commented on main/src/test/scala/ParseKey.scala:
```diff
@@ -22,13 +22,22 @@ object ParseKey extends Properties("Key parser test") {
+        mask0 match {
+          case _ if hasZeroConfig => mask0.copy(project = true)
+          case _                  => mask0
+        }
```
this is overly complex..

```scala
val mask = if (hasZeroConfig) mask0.copy(project = true) else mask0
```

dwijnand commented on main/src/test/scala/ParseKey.scala:
```diff
@@ -22,13 +22,22 @@ object ParseKey extends Properties("Key parser test") {
+      // This is to disambiguate `proj/Zero/name`, which could render potentially
+      // as `Zero/name`, but could be interpretted as `Zero/Zero/name`.
+      val s = displayMasked(key, mask, hasZeroConfig)
+      // println(s"s: $s")
```
remove

dwijnand commented on main/src/test/scala/ParseKey.scala:
```diff
@@ -101,7 +116,10 @@ object ParseKey extends Properties("Key parser test") {
-      case Right(sk) => Project.equal(sk, expected, mask)
+      case Right(sk) =>
+        // println(s"${sk}.key == ${expected}.key: ${sk.key == expected.key}")
+        // println(s"${sk.scope} == ${expected.scope}: ${Scope.equal(sk.scope, expected.scope, mask)}")
```
remove

eed3si9n replied:
Actually the right thing to do is convert it into ScalaCheck label so it'll be displayed when it fails:

```scala
        (s"${sk}.key == ${expected}.key: ${sk.key == expected.key}") |:
          (s"${sk.scope} == ${expected.scope}: ${Scope.equal(sk.scope, expected.scope, mask)}") |:
          Project.equal(sk, expected, mask)
```

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

dwijnand commented on main-settings/src/main/scala/sbt/Scope.scala:
```diff
@@ -123,29 +123,84 @@ object Scope {
+    displayMasked(scope, sep, showProject, mask, false)
+
+  /**
+   * unified slash style introduced in sbt sbt 1.1.0
```
sbt sbt

dwijnand reviewed (changes requested):
2 requests, 1 more typo

eed3si9n committed:
Add whitespaces in Show of scoped keys

```
Provided by:
    ProjectRef(uri("...."), "root") / Test / test
Dependencies:
    Test / executeTests
    Test / test / streams
    Test / state
    Test / test / testResultLogger
```

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>
